### PR TITLE
python3Packages.scikitimage: 0.18.1 -> 0.18.3

### PR DESCRIPTION
**Pull request by yl3dy**
###### Motivation for this change

Version bump + added missing tests. Also added myself as a maintainer, hope it's ok.

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [x] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/#sec-conf-file))
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [21.11 Release Notes (or backporting 21.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2111-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by yl3dy** [comment]
Result of `nixpkgs-review pr 137038` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>2 packages marked as broken and skipped:</summary>
  <ul>
    <li>python38Packages.rl-coach</li>
    <li>python39Packages.rl-coach</li>
  </ul>
</details>
<details>
  <summary>40 packages built:</summary>
  <ul>
    <li>python38Packages.aplpy</li>
    <li>python38Packages.asdf</li>
    <li>python38Packages.batchgenerators</li>
    <li>python38Packages.caffe</li>
    <li>python38Packages.dask-image</li>
    <li>python38Packages.glymur</li>
    <li>python38Packages.image-match</li>
    <li>python38Packages.imagecorruptions</li>
    <li>python38Packages.imgaug</li>
    <li>python38Packages.mask-rcnn</li>
    <li>python38Packages.pims</li>
    <li>python38Packages.pyregion</li>
    <li>python38Packages.pytest-astropy</li>
    <li>python38Packages.pytest-astropy-header</li>
    <li>python38Packages.reproject</li>
    <li>python38Packages.scikitimage</li>
    <li>python38Packages.slicedimage</li>
    <li>python38Packages.spectral-cube</li>
    <li>python38Packages.stytra</li>
    <li>python38Packages.sunpy</li>
    <li>python39Packages.aplpy</li>
    <li>python39Packages.asdf</li>
    <li>python39Packages.batchgenerators</li>
    <li>python39Packages.caffe</li>
    <li>python39Packages.dask-image</li>
    <li>python39Packages.glymur</li>
    <li>python39Packages.image-match</li>
    <li>python39Packages.imagecorruptions</li>
    <li>python39Packages.imgaug</li>
    <li>python39Packages.mask-rcnn</li>
    <li>python39Packages.pims</li>
    <li>python39Packages.pyregion</li>
    <li>python39Packages.pytest-astropy</li>
    <li>python39Packages.pytest-astropy-header</li>
    <li>python39Packages.reproject</li>
    <li>python39Packages.scikitimage</li>
    <li>python39Packages.slicedimage</li>
    <li>python39Packages.spectral-cube</li>
    <li>python39Packages.stytra</li>
    <li>python39Packages.sunpy</li>
  </ul>
</details>

**Comment by r-rmcgibbo** [comment]
Result of `nixpkgs-review pr 137038` at 1325c11b run on aarch64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>2 packages marked as broken and skipped:</summary>
  <ul>
    <li>python38Packages.caffe</li>
    <li>python39Packages.caffe</li>
  </ul>
</details>
<details>
  <summary>40 packages failed to build:</summary>
  <ul>
    <li><a href="https://gist.github.com/8a3c66d22e8f4aaac9cdd4497f62ff51">python38Packages.aplpy</a></li>
    <li><a href="https://gist.github.com/129f31d612919e46eafb08f59a3b5bea">python38Packages.asdf</a></li>
    <li><a href="https://gist.github.com/ddaf26d0afe5639a6ce4aaab4e8cad52">python38Packages.batchgenerators</a></li>
    <li><a href="https://gist.github.com/29bc29056b8a44ad4b95f57d514baa9c">python38Packages.dask-image</a></li>
    <li><a href="https://gist.github.com/dca6ef3e709ba61bc713811602734156">python38Packages.glymur</a></li>
    <li><a href="https://gist.github.com/5504bab5ba6cb2a5f3c2206dce5389be">python38Packages.image-match</a></li>
    <li><a href="https://gist.github.com/510d5b8a14aab5089ebce0aa7e45892e">python38Packages.imagecorruptions</a></li>
    <li><a href="https://gist.github.com/f9fc7e3c484eed82daf56bcbac645ce6">python38Packages.imgaug</a></li>
    <li><a href="https://gist.github.com/4c62e9a2fcfde945952ef6bc71deaafe">python38Packages.mask-rcnn</a></li>
    <li><a href="https://gist.github.com/89c49892229fe940abd329662037a77f">python38Packages.pims</a></li>
    <li><a href="https://gist.github.com/ce4e7af13f2db1f409c2728da3b08112">python38Packages.pyregion</a></li>
    <li><a href="https://gist.github.com/cc6b01b22caccad6eceeb39308778169">python38Packages.pytest-astropy</a></li>
    <li><a href="https://gist.github.com/1351e94d72922c99c83fb558d4dbfd74">python38Packages.pytest-astropy-header</a></li>
    <li><a href="https://gist.github.com/7a71fd2e6a29f173c71af9d847d73fb2">python38Packages.reproject</a></li>
    <li><a href="https://gist.github.com/e6e174d613d5c9e2da5d1efe1079a936">python38Packages.rl-coach</a></li>
    <li><a href="https://gist.github.com/53022f7b372336d91371cbbffdc9b7cb">python38Packages.scikitimage</a></li>
    <li><a href="https://gist.github.com/b706c93825ecf0526965db64bba2356e">python38Packages.slicedimage</a></li>
    <li><a href="https://gist.github.com/01b8b2e6fe56cfcc689c74a57ba30508">python38Packages.spectral-cube</a></li>
    <li><a href="https://gist.github.com/2b8e63e6dedff47cb850b478dbff8d4f">python38Packages.stytra</a></li>
    <li><a href="https://gist.github.com/f2d220b5c73bd335901c222dc40783f6">python38Packages.sunpy</a></li>
    <li><a href="https://gist.github.com/733a7f9fc8371387054ed61dfa371360">python39Packages.aplpy</a></li>
    <li><a href="https://gist.github.com/e583519dd2fd2ea21ddbaafc594aabec">python39Packages.asdf</a></li>
    <li><a href="https://gist.github.com/c01ba3a68d8cd2e90c39a0b9f88bc5a7">python39Packages.batchgenerators</a></li>
    <li><a href="https://gist.github.com/fee643d4f8cf25dd769e8f8747bce4dd">python39Packages.dask-image</a></li>
    <li><a href="https://gist.github.com/3871107ea606ae981d775900d229e5a3">python39Packages.glymur</a></li>
    <li><a href="https://gist.github.com/e2a4fdce1813125f9c53916f9b072b0d">python39Packages.image-match</a></li>
    <li><a href="https://gist.github.com/71f29d912fb715175c5c9451203bd2e4">python39Packages.imagecorruptions</a></li>
    <li><a href="https://gist.github.com/daae4e80395fe5273957a4c54ea56396">python39Packages.imgaug</a></li>
    <li><a href="https://gist.github.com/86133a2190b4fccf2453ab729b6c25ba">python39Packages.mask-rcnn</a></li>
    <li><a href="https://gist.github.com/c82c1d64695e7a5a50fbf740d0923bb0">python39Packages.pims</a></li>
    <li><a href="https://gist.github.com/25deca6c6f5ec441f201a57fb94902f2">python39Packages.pyregion</a></li>
    <li><a href="https://gist.github.com/8f53dd47fb8b4d1311483b81f4f2c8c7">python39Packages.pytest-astropy</a></li>
    <li><a href="https://gist.github.com/78e4bdb81d0c4c42bafdf4af3c062718">python39Packages.pytest-astropy-header</a></li>
    <li><a href="https://gist.github.com/a1f6821909a769af7e4cbfde5ed28833">python39Packages.reproject</a></li>
    <li><a href="https://gist.github.com/ed2a0590d159fcdafd91e085ba17e8db">python39Packages.rl-coach</a></li>
    <li><a href="https://gist.github.com/442349fc65a4d423f77b474360ff1dda">python39Packages.scikitimage</a></li>
    <li><a href="https://gist.github.com/1ef70bd2d0a3c6e05d88522a45801622">python39Packages.slicedimage</a></li>
    <li><a href="https://gist.github.com/504720be7171d0ff3f1317c361a816d8">python39Packages.spectral-cube</a></li>
    <li><a href="https://gist.github.com/d50a9f621a692b3562edaf2506d4eca5">python39Packages.stytra</a></li>
    <li><a href="https://gist.github.com/f4067a06794dde1dbb728f26aa79bbe3">python39Packages.sunpy</a></li>
  </ul>
</details>

Note that build failures [may predate](https://github.com/nix-community/hydra-check) this PR, and could be nondeterministic or hardware dependent.
Please exercise your independent judgement. Does something look off? Please file an [issue](https://github.com/rmcgibbo/r-rmcgibbo/issues) or reach out on IRC.

**Comment by risicle** [comment]
This results in an increased output size from 33M to 43M.

File: pkgs/development/python-modules/scikit-image/default.nix [hunk 1]
```diff
@@ -43,12 +52,41 @@ buildPythonPackage rec {
...
   checkInputs = [ pytest ];
 
-  # No tests in archive
-  doCheck = false;
+  # See the explanation above
+  checkPhase = ''
+    rm -r skimage
+    export INSTALL_PATH=$out/${python.sitePackages}
+    pytest $INSTALL_PATH \
+      --ignore=$INSTALL_PATH/skimage/filters/rank/tests/test_rank.py \
```
Comment (risicle) [inline pkgs/development/python-modules/scikit-image/default.nix:60]: It would be good to include reasons for the disabled tests, otherwise it just looks like we're disabling the ones that don't pass, which rather defeats the point ;)

File: pkgs/development/python-modules/scikit-image/default.nix [hunk 2]
```diff
@@ -11,20 +12,28 @@
...
+    owner = pname;
+    repo = pname;
+    rev = "v0.18.3";
+    sha256 = "0a2h3bw5rkk23k4r04qc9maccg00nddssd7lfsps8nhp5agk1vyh";
   };
 
+  # (1) The package has cythonized modules, whose .so libs will appear only in the wheel, i.e. in nix store;
+  # (2) To stop Python from importing the wrong directory, i.e. the one in the build dir, not the one in nix store, `skimage` dir should be (re)moved;
+  # (3) Therefore, tests should be run on the nix-store version.
+  # Because of that, a custom checkPhase is needed along with the patch to add all the data/image files for testing, not only the "legacy" set.
```
Comment (samuela) [inline pkgs/development/python-modules/scikit-image/default.nix:34]: This doesn't make sense to me... What exactly is wrong?

File: pkgs/development/python-modules/scikit-image/default.nix [hunk 3]
```diff
@@ -43,12 +52,41 @@ buildPythonPackage rec {
 
   checkInputs = [ pytest ];
 
-  # No tests in archive
-  doCheck = false;
+  # See the explanation above
+  checkPhase = ''
+    rm -r skimage
+    export INSTALL_PATH=$out/${python.sitePackages}
+    pytest $INSTALL_PATH \
```
Comment (samuela) [inline pkgs/development/python-modules/scikit-image/default.nix:59]: Is there a good reason not to use pytestCheckHook? If so, I think that deserves explanation in a comment.

File: pkgs/development/python-modules/scikit-image/default.nix [hunk 4]
```diff
@@ -43,12 +52,41 @@ buildPythonPackage rec {
 
   checkInputs = [ pytest ];
 
-  # No tests in archive
-  doCheck = false;
+  # See the explanation above
+  checkPhase = ''
+    rm -r skimage
+    export INSTALL_PATH=$out/${python.sitePackages}
```
Comment (samuela) [inline pkgs/development/python-modules/scikit-image/default.nix:58]: Why is this necessary?

**Review by samuela — CHANGES_REQUESTED** [review]
I'm confused as to why the patch and custom checkPhase are necessary.

File: pkgs/development/python-modules/scikit-image/default.nix [hunk 5]
```diff
@@ -11,20 +12,28 @@
...
-  version = "0.18.1";
+  version = "0.18.3";
 
-  src = fetchPypi {
-    inherit pname version;
-    sha256 = "fbb618ca911867bce45574c1639618cdfb5d94e207432b19bc19563d80d2f171";
+  src = fetchFromGitHub {
+    owner = pname;
+    repo = pname;
+    rev = "v0.18.3";
```
Comment (SuperSandro2000) [inline pkgs/development/python-modules/scikit-image/default.nix:27]:
```suggestion
    rev = "v${version}";
```

File: pkgs/development/python-modules/scikit-image/add_testing_data.patch [hunk 6]
```diff
@@ -0,0 +1,18 @@
+diff --git a/skimage/data/setup.py b/skimage/data/setup.py
+index 528e9c284ce..bfc5ad97d0b 100644
+--- a/skimage/data/setup.py
++++ b/skimage/data/setup.py
+@@ -11,7 +11,12 @@ def configuration(parent_package='', top_path=None):
+     # further notice.
+     # Testing data and additional datasets should only
+     # be made available via pooch
+-    config.add_data_files(*legacy_datasets)
++    #config.add_data_files(*legacy_datasets)
```
Comment (SuperSandro2000) [inline pkgs/development/python-modules/scikit-image/add_testing_data.patch:10]:
```suggestion
```

**Review by SuperSandro2000 — CHANGES_REQUESTED** [review]

File: pkgs/development/python-modules/scikit-image/default.nix (hunk 1 above)
Comment (yl3dy) [reply pkgs/development/python-modules/scikit-image/default.nix:60]: Of course, have forgot about it. These tests require network connection, will add a comment.

File: pkgs/development/python-modules/scikit-image/default.nix (hunk 2 above)
Comment (yl3dy) [reply pkgs/development/python-modules/scikit-image/default.nix:34]: Please see https://discourse.nixos.org/t/cant-import-cythonized-modules-at-checkphase/14207 about the problem with cythonized modules and checks.

File: pkgs/development/python-modules/scikit-image/default.nix (hunk 3 above)
Comment (yl3dy) [reply pkgs/development/python-modules/scikit-image/default.nix:59]: The reason is to add correct paths to tests, both the root test path and ignore paths. I will add it to the comments.

File: pkgs/development/python-modules/scikit-image/default.nix (hunk 5 above)
Comment (yl3dy) [reply pkgs/development/python-modules/scikit-image/default.nix:27]: Fair enough, thanks!

File: pkgs/development/python-modules/scikit-image/add_testing_data.patch (hunk 6 above)
Comment (yl3dy) [reply pkgs/development/python-modules/scikit-image/add_testing_data.patch:10]: Thanks!

File: pkgs/development/python-modules/scikit-image/default.nix (hunk 4 above)
Comment (yl3dy) [reply pkgs/development/python-modules/scikit-image/default.nix:58]: `INSTALL_PATH` is just a helper variable to form paths to test modules in the nix store. It points to the installed package path.

File: pkgs/development/python-modules/scikit-image/default.nix (hunk 3 above)
Comment (SuperSandro2000) [reply pkgs/development/python-modules/scikit-image/default.nix:67]: We should be able to add the argument with pytestFlagsArray and use the other flags to ignore the failing tests.

Comment (yl3dy) [reply pkgs/development/python-modules/scikit-image/default.nix:67]: Could it be made when a runtime value is needed? I mean, the `$out/${python.sitePackages}` is known only at runtime, which is needed to specify e.g. test root.

**Comment by risicle** [comment]
Still uncertain whether we need to be shipping 16M of image files in this package (see `lib/python3.8/site-packages/skimage/data`)

**Comment by risicle** [comment]
Builds successfully on macos 10.15 too.

File: pkgs/development/python-modules/scikit-image/default.nix (hunk 2 above)
Comment (samuela) [reply pkgs/development/python-modules/scikit-image/default.nix]: It would be good to add that link in the comments so that future code readers have a paper trail for this.

**Comment by samuela** [comment]
> Still uncertain whether we need to be shipping 16M of image files in this package (see `lib/python3.8/site-packages/skimage/data`)
> 
> Though, of course, I always applaud efforts at enabling tests.

Could the images just be removed in `postCheck`?

**Comment by yl3dy** [comment]
> Still uncertain whether we need to be shipping 16M of image files in this package (see `lib/python3.8/site-packages/skimage/data`)
> 
> Though, of course, I always applaud efforts at enabling tests.

Well, there's tradeoff between testing as much as possible and carrying extra images. My reasoning is that the package is most often used in research environment, where additional 16 MB is not a big deal, so it's better to test more.

**Comment by yl3dy** [comment]
> > Still uncertain whether we need to be shipping 16M of image files in this package (see `lib/python3.8/site-packages/skimage/data`)
> > Though, of course, I always applaud efforts at enabling tests.
> 
> Could the image files just be removed in `postCheck`?

Not really, these images are a part of wheel and get installed into nix store. Don't think we can (or should) modify data there after install.

File: pkgs/development/python-modules/scikit-image/default.nix (hunk 2 above)
Comment (yl3dy) [reply pkgs/development/python-modules/scikit-image/default.nix]: It seems that a number of existing packages do use a similar trick, but your point is valid, of course.

File: pkgs/development/python-modules/scikit-image/default.nix (hunk 3 above)
Comment (SuperSandro2000) [reply pkgs/development/python-modules/scikit-image/default.nix:67]: > Could it be made when a runtime value is needed?

I don't understand your question. We can add arbitrary things to pytest.

Comment (risicle) [reply pkgs/development/python-modules/scikit-image/default.nix:67]: But `$out` is only readable as an environment variable (i.e. only known at runtime, whereas most of the values we pass to `pytestCheckHook` are baked into the `.drv`)

Comment (SuperSandro2000) [reply pkgs/development/python-modules/scikit-image/default.nix:67]: ``${placeholder $out}``?

Comment (yl3dy) [reply pkgs/development/python-modules/scikit-image/default.nix:67]: > `${placeholder $out}`?

Sorry, never saw this feature in the manual. I will update the expression accordingly.

File: pkgs/development/python-modules/scikit-image/default.nix [hunk 7]
```diff
@@ -44,35 +48,29 @@ buildPythonPackage rec {
...
-    pytest $INSTALL_PATH \
-      --ignore=$INSTALL_PATH/skimage/filters/rank/tests/test_rank.py \
-      --deselect=skimage/data/test_data.py::test_skin \
-      --deselect=skimage/data/tests/test_data.py::test_skin \
-      --deselect=skimage/io/tests/test_io.py::test_imread_http_url \
-      --deselect=skimage/restoration/tests/test_rolling_ball.py::test_ndim \
-      --pyargs skimage
   '';
 
+  disabledTestPaths = [
```
Comment (samuela) [inline pkgs/development/python-modules/scikit-image/default.nix:62]: Didn't know about the disabledTestPaths option... neat!

File: pkgs/development/python-modules/scikit-image/default.nix [hunk 8]
```diff
@@ -41,14 +48,51 @@ buildPythonPackage rec {
...
+  # (1) The package has cythonized modules, whose .so libs will appear only in the wheel, i.e. in nix store;
+  # (2) To stop Python from importing the wrong directory, i.e. the one in the build dir, not the one in nix store, `skimage` dir should be removed or renamed;
+  # (3) Therefore, tests should be run on the installed package in nix store.
 
-  # No tests in archive
-  doCheck = false;
+  # See e.g. https://discourse.nixos.org/t/cant-import-cythonized-modules-at-checkphase/14207 on why the following is needed.
+  preCheck = ''
+    rm -r skimage
+  '';
```
Comment (samuela) [inline pkgs/development/python-modules/scikit-image/default.nix:60]: I just packaged another cython package, and didn't need this in the end. [Commit here](https://github.com/samuela/nixpkgs/commit/8fda9a49a7b0bf6159d2bb46c64022cffe59dce0). Have you tried removing this now that pytest is being pointed towards `installedPackageRoot`?

Comment (yl3dy) [reply pkgs/development/python-modules/scikit-image/default.nix:60]: Well, in this case it still breaks pytest, specifically test collection:
```
E   _pytest.pathlib.ImportPathMismatchError: ('skimage.conftest', '/nix/store/sj4bdpny8w0kh7dqvgbnz7asgi7rjg93-python3.8-scikit-image-0.18.3/lib/python3.8/site-packages/skimage/conftest.py', PosixPath('/build/source/skimage/conftest.py'))
```

I think it depends on an actual test configuration. Here it's probably easier to leave it as is.

**Comment by SuperSandro2000** [comment]
This is a semi-automatic executed nixpkgs-review with [nixpkgs-review-checks extension](https://github.com/SuperSandro2000/nixpkgs-review-checks). It is checked by a human on a best effort basis and does not build all packages (e.g. lumo, tensorflow or pytorch).

Result of `nixpkgs-review pr 137038` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
<summary>
2 packages marked as broken and skipped:
</summary>
<ul>
<li>python38Packages.rl-coach</li>
<li>python39Packages.rl-coach</li>
</ul>
</details>
<details>

  <summary>2 packages failed to build and already failed to build on hydra master:</summary>
  <ul>
<li>python38Packages.spectral-cube: log was empty</li>
<li>python39Packages.spectral-cube: log was empty</li>
  </ul>
</details>
<details>
<summary>
38 packages built:
</summary>
<ul>
<li>intensity-normalization (python39Packages.intensity-normalization)</li>
<li>python38Packages.aplpy</li>
<li>python38Packages.asdf</li>
<li>python38Packages.batchgenerators</li>
<li>python38Packages.caffe</li>
<li>python38Packages.dask-image</li>
<li>python38Packages.glymur</li>
<li>python38Packages.image-match</li>
<li>python38Packages.imagecorruptions</li>
<li>python38Packages.imgaug</li>
<li>python38Packages.intensity-normalization</li>
<li>python38Packages.pims</li>
<li>python38Packages.pyregion</li>
<li>python38Packages.pytest-astropy</li>
<li>python38Packages.pytest-astropy-header</li>
<li>python38Packages.reproject</li>
<li>python38Packages.scikitimage</li>
<li>python38Packages.slicedimage</li>
<li>python38Packages.stytra</li>
<li>python38Packages.sunpy</li>
<li>python39Packages.aplpy</li>
<li>python39Packages.asdf</li>
<li>python39Packages.batchgenerators</li>
<li>python39Packages.caffe</li>
<li>python39Packages.dask-image</li>
<li>python39Packages.glymur</li>
<li>python39Packages.image-match</li>
<li>python39Packages.imagecorruptions</li>
<li>python39Packages.imgaug</li>
<li>python39Packages.pims</li>
<li>python39Packages.pyregion</li>
<li>python39Packages.pytest-astropy</li>
<li>python39Packages.pytest-astropy-header</li>
<li>python39Packages.reproject</li>
<li>python39Packages.scikitimage</li>
<li>python39Packages.slicedimage</li>
<li>python39Packages.stytra</li>
<li>python39Packages.sunpy</li>
</ul>
</details>
